### ISSUE TITLE
False negative for `curly_braces_in_control_flow_structures`

### ISSUE THREAD

**Issue by lrhn** [dart-lang/linter#4870]
The lint [`curly_braces_in_control_flow_structures`](https://dart.dev/tools/linter-rules/curly_braces_in_flow_control_structures) should warn if an `if` statement body has no braces. 
An exception is granted if the *entire `if` statement* fits in one line, based on the corresponding [Effective Dart rule](https://dart.dev/effective-dart/style#do-use-curly-braces-for-all-flow-control-structures).

No lint warning is given for the `if` statement of this code:
```dart
Object? foo() {  
  if ("long condition and other long condition" != "another long condition" &&
     "more tests".isNotEmpty) return null;
  return 42;
}
```

The body of this `if` statement fits in one line, but the entire statement does not, so the lint should trigger.
It does not.

(Found this by seeing similar code in a code review.)

**Comment by navaronbracke** [comment]
But what if someone wants to enforce that curly_braces_in_control_flow_structures enforces _all_ control flow to have braces?

IMO, the whole point of the lint was to avoid `if (condition) return;` ? Can we configure this?

**Comment by srawlins** [comment]
> But what if someone wants to enforce that curly_braces_in_control_flow_structures enforces all control flow to have braces?

I don't think that's related to this issue...

**Comment by navaronbracke** [comment]
With the change proposal above, the statement `if (condition) return;` would not be flagged by the lint, per the rule in the style guide. However, I propose to keep the current behavior and lint on `if (condition) return;` _regardless_ of the line length for the full statement.

**Comment by srawlins** [comment]
`if (condition) return;` is not flagged today though. The change proposal above would not change that.

**Comment by navaronbracke** [comment]
Ah, my mistake. Should I file a new issue for having it linted?

**Comment by srawlins** [comment]
Yes please :D.

Without a backing Effective Dart rule though, we are unlikely to add a style-oriented rule, or change the Effective Dart rule. You can file a request to change the style guide at https://github.com/dart-lang/site-www/issues.

**Comment by lrhn** [comment]
Agree. The issue here is that the expected exception seems to be implemented by checking that the `)` ending the condition and the end of the statement are on the same line. 
It should instead check that the `if` and the end of the statement is on the same line.
(I checked the code, that's precisely what happens. Fix seems easy, so I [wrote that too](https://dart-review.googlesource.com/c/sdk/+/353741).

**Comment by srawlins** [comment]
Sorry, forgot to mention I've been working on this and cleaning up the SDK. https://dart-review.googlesource.com/c/sdk/+/353140

But it will take a lot of work to land. Google internal needs a lot of cleanup.

**Comment by lrhn** [comment]
SGTM, ship it! (But yes, signficant clean-up will likely be needed.)

Feel free to take the [extra tests](https://dart-review.googlesource.com/c/sdk/+/353741/1/pkg/linter/test/rules/curly_braces_in_flow_control_structures_test.dart#243) I added, if you think they can be useful.

**Comment by srawlins** [comment]
Fixed by https://github.com/FMorschel/sdk/commit/2c6dec625ec4bb2b540bfabcb1922d460e72836c

**Comment by FMorschel** [comment]
Just to be clear, I've nothing to do with the fix. I'm not sure why, but some of the time when I rebase my fork on GH it keeps warning issues about changes from other people but it is simply because the link is commented on the commit description. Sorry for the confusion!
